Fix for when recommendations are short

saaalik opened the pull request:
## Description
- Fixed `get_rated_bookids` function
- When combining the two recommendation lists, if that list<`recommendations`(9) then my code snippet is executed which fills in remaining books to be recommended, half from cv_bookids and the other half with top rated books of the most common genre among `already_rated_books` and the books yet recommended.
- To recommend most rated genre book, A pickle file was created to store the value of the dictionary storing the same, files used can be accessed [here](https://easyupload.io/m/m87iol)
- Issue #38 

## Affected Dependencies
Changes are made only into `mainapp/views.py` and `helper.py` affecting the personalized-recommendations section of the project

## How has this been tested?
Tested on python shell manually as well as on the main project webpage. Rated 5 different books, when the `best_bookids` was calculated manually, only 5 books were recommended even after summing up bookids from `cv_bookids` and `embedding_bookids` and thus my code snippet executed and added 2 bookids from `cv_bookids` and 2 similar genre books

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/Praful932/Kitabe/blob/master/CONTRIBUTING.md).
- [x] My changes do not edit/add any unrequired files.
- [x] My changes are covered by tests.

Praful932 commented on mainapp/helpers.py:
```diff
@@ -19,6 +22,9 @@
 sim_books_path = os.path.join(settings.STATICFILES_DIRS[0] + '/mainapp/model_files/surprise/sim_books.pickle')
 
+# For Similar-Genre books
+sim_genre_books_path = os.path.join(settings.STATICFILES_DIRS[0] + '/mainapp/model_files/genre_dictionary.pickle')
```
Could you explain, Why is this needed? 
The remaining `n` 's genre as discussed should be taken from the user's most common(rated & recommended) genre and return n top books of that genre not present in `recommended & rated`, I think `df_book` already has all what you need

saaalik replied:
The file `genre_dictionary.pickle` contains inverted index of all different genres, sorted in descending order of votes from column `count` in `full_genre.csv`, making the process less time consuming, else the script would have to loop through a lot of books checking for the highest rating and the corresponding genre. If `count` column (genre votes) is not a directly proportional variable to the book_ratings then I can make an inverted index using books.csv such that
```python
genre_index = {genre: [ [rating, book_id], [rating2, book_id2] ] ,
                         genre2: [ [rating, book_id],[rating, book_id],[rating, book_id] ],
                         genre3: [].......}
```
sorted in descending order of rating. Then directly importing it to the function this variable is being used in.

Praful932 replied:
Better reuse [this](https://github.com/Praful932/Kitabe/blob/master/mainapp/helpers.py#L98)function, add a parameter `n_books`(default = 16) - For genre page.
Use this function to get total `2*needed_books` (to avoid duplicates)  for any particular genre

Praful932 commented on mainapp/views.py:
```diff
@@ -58,7 +58,7 @@ def book_recommendations(request):
     best_user_ratings = sorted(user_ratings, key=operator.attrgetter('bookrating'), reverse=True)
 
-    if len(best_user_ratings) < 4:
+    if len(best_user_ratings) < 5:
```
You need to change this back, even if user has `5` ratings it will not pass

saaalik replied:
Okey, I will revert back this change

Praful932 commented:
@saaalik Re-request a review when the comments are resolved, Thanks

Praful932 commented on mainapp/helpers.py:
```diff
@@ -95,12 +96,12 @@ def get_bookid(raw_id_list):
     '''
-    n_books = 16
     min_genre_book_count = 48
+    cols = ['average_rating', 'book_id']
```
This will break `genre` page

Praful932 commented on mainapp/helpers.py:
```diff
@@ -112,7 +113,7 @@ def genre_wise(genre, percentile=0.85):
     qualified.sort_values('weighted_rating', ascending=False, inplace=True)
 
-    return qualified[cols].head(min_genre_book_count).sample(n_books)
+    return qualified[cols].head(min_genre_book_count).sample(int(2*n_books))
```
why `int`, let's pass an already converted `integer` value instead of leaving it to the function

Praful932 commented on mainapp/helpers.py:
```diff
@@ -186,10 +187,34 @@ def combine_ids(cv_bookids, embedding_bookids, already_rated, recommendations=9)
-        best_bookids = best_bookids + cv_bookids[3:(recommendations + 3 - len(best_bookids))]
+        two_n = (recommendations - len(best_bookids))
+        n1, n2 = math.ceil(two_n/2), math.floor(two_n/2)
+        best_bookids_remain = cv_bookids[3: 6+n1]
```
If you put `6` because of duplicates, Specify `3 * 2`, for more clarity, +1 if you add comments wherever possible, since for a new contributor it's a bit difficult to catch up

Praful932 commented on mainapp/helpers.py:
```diff
@@ -186,10 +187,34 @@ def combine_ids(cv_bookids, embedding_bookids, already_rated, recommendations=9)
+        Returns n top rated of the most_common_genre among all lists taken as input
+    '''
+    # Final list of bookids to be recommended
+    books = set(best_bookids+list(already_rated))
```
You need to find common genre among -  `best_bookids`,`already_rated` & the first half which will be recommended

Praful932 reviewed (changes requested):
I have added some comments, resolve them. Also it would be great if you replace wherever `cv` method or variable is mentioned to `tfidf` since we adopted the tfidf approach now.
And rename `best_bookids_remain` to `best_bookids_tfidf` since they are the first half of the remaining `2n` to be recommended, revert any line changes such as `32` in `helpers.py` and remove the pickle file, since we aren't using that now.

Praful932 commented on mainapp/helpers.py:
```diff
@@ -95,11 +96,12 @@ def get_bookid(raw_id_list):
     '''
-    n_books = 16
+    # To return double the number of books required
+    n_books = n_books * 2
```
😅, What I meant is keep this as it is, whenver you pass, pass it `x2`

Praful932 commented on mainapp/helpers.py:
```diff
@@ -170,26 +172,63 @@ def get_book_dict(bookid_list):
+
+    # The most common genre among the bookids in `books` variable
+    most_common_genre = Counter(genre_frequency).most_common(1)[0][0]
+
```
@saaalik I could follow till here, you got the most common genre, could you elaborate on the code below and why is there a loop needed?

saaalik replied:
genre_wise function will always return 2*n number of books, but if a case came up where the unique sets of bookids are less than n after removing the common bookids from `books` i.e. after 
```python
genre_recommendations = genre_recommendations.difference(books)    #Might return a list of length<n in rare test cases#
```, then the genre_wise function will be executed again, and the weight calculation method always randomizes the output, so the loop will execute till the returned value of genre_wise function is less than n

Praful932 replied:
`books`  will always have a unique set since we took care of that, so whatever remaining `n` books we take from whatever genre, just make it `2*n` and take a `difference` and take first `n` books, don't see why loop would be needed

saaalik replied:
Removed the while loop, please review

Praful932 replied:
Why is the code below changed after you finished with the implementation earlier and were about to write the tests, it was correct then

Praful932 reviewed (changes requested):
Delete the extra `pickle` file

Praful932 commented on mainapp/helpers.py:
```diff
@@ -112,7 +112,7 @@ def genre_wise(genre, percentile=0.85):
     qualified.sort_values('weighted_rating', ascending=False, inplace=True)
 
-    return qualified[cols].head(min_genre_book_count).sample(n_books)
+    return qualified[cols].head(min_genre_book_count).sample(2*n_books)
```
If you modify this, then `2*n` goes to genre too, pass `2*n` from where you need itself

Praful932 commented on mainapp/helpers.py:
```diff
@@ -170,26 +170,61 @@ def get_book_dict(bookid_list):
+
+    # Recommendations list, listing n or more unique bookids
+    genre_recommendations = set()
+    genre_recommendations = set(genre_wise(most_common_genre, n).book_id.to_numpy())
```
`2*n` here

saaalik replied:
ooh alright, I'll do that

Praful932 reviewed (changes requested):
Added some comments, also you would need to write unit tests for the `most_common_genre_recommendations` function, +1 if you change the  `count_vectorizer_recommendations` function name to `tfidf_recommendations`

saaalik commented:
> Added some comments, also you would need to write unit tests for the `most_common_genre_recommendations` function, +1 if you change the `count_vectorizer_recommendations` function name to `tfidf_recommendations`

I don't know how what books will be genreated by genre_wise function for any test cases, because genre_wise function returns different books every time, What I can do is check for the most_common_genre among books i.e.(`set(best_bookids + list(already_rated) + best_bookids_tfidf)`) and cross check the genre of n2 recommended books by the function `most_common_genre_recommendations` like [this...](https://easyupload.io/omoq0k)
In this test case the most common genre is `fiction`

Praful932 commented:
> > Added some comments, also you would need to write unit tests for the `most_common_genre_recommendations` function, +1 if you change the `count_vectorizer_recommendations` function name to `tfidf_recommendations`
> 
> I don't know how what books will be genreated by genre_wise function for any test cases, because genre_wise function returns different books every time, What I can do is check for the most_common_genre among books i.e.(`set(best_bookids + list(already_rated) + best_bookids_tfidf)`) and cross check the genre of n2 recommended books by the function `most_common_genre_recommendations` like [this...](https://easyupload.io/omoq0k)
> In this test case the most common genre is `fiction`
Yes exactly you can do that and vary the test cases with edge ones

saaalik commented:
[These](https://easyupload.io/m/9he5yl) are the test cases for the function, please have a look

Praful932 reviewed (changes requested):
Unstage `settings.py`, also post the ss here, it would be easier for us to review, you can directly `ctrl + v` in the PR after taking a snip `ctrl + shift + s` :)

Praful932 commented:
A general guideline for creating TC for your changes
- Assume `already`  has atleast >= 5 books
- `bestbookids` - embeddings can have 0-6 books
- All other variables will change according to the above variables, you already know what values will they be if you fix some values above, so you just have to assert that, similarly make multiple test cases for different scenarios

saaalik commented:
### TESTCASES for `most_common_genre_recommendations`

![Prerequisites](https://user-images.githubusercontent.com/65087495/113993885-87626b80-9872-11eb-8900-e891b3b910d9.png)
![testcase1](https://user-images.githubusercontent.com/65087495/113993968-9ba66880-9872-11eb-9773-dbfe2fc0697f.png)
![testcase2](https://user-images.githubusercontent.com/65087495/113994027-a95bee00-9872-11eb-88f9-e710faa7b625.png)
![testcase3](https://user-images.githubusercontent.com/65087495/113994041-abbe4800-9872-11eb-87cb-d66447b5b925.png)
![testcase4](https://user-images.githubusercontent.com/65087495/113994047-ae20a200-9872-11eb-97e3-c19e4002c3b5.png)
![testcase5](https://user-images.githubusercontent.com/65087495/113994056-afea6580-9872-11eb-8791-daea2a247b57.png)
![testcase6](https://user-images.githubusercontent.com/65087495/113994064-b24cbf80-9872-11eb-88af-5222ea2f5dc5.png)
![testcase7](https://user-images.githubusercontent.com/65087495/113994088-b5e04680-9872-11eb-96f8-a78d2846a48f.png)
![testcase8](https://user-images.githubusercontent.com/65087495/113994097-b8db3700-9872-11eb-910c-ac4928a882ad.png)

### TESTCASES for the Counter bit edited to return most common genre based on priority list
[Link](https://easyupload.io/61mj2x) to the py file for testing
![Screenshot (259)](https://user-images.githubusercontent.com/65087495/113994720-3dc65080-9873-11eb-876c-d9874e9f2882.png)
![Screenshot (260)](https://user-images.githubusercontent.com/65087495/113994733-40c14100-9873-11eb-8bc8-933cbce52358.png)

Praful932 reviewed (changes requested):
I don't see any changes in test files? You need to add the tests in that file and the `most_common_genre_recommendations` seems to be changed? It was fine..

saaalik commented:
Is testfiles a folder in the repo? I don't understand

Praful932 commented:
@saaalik [This one](https://github.com/Praful932/Kitabe/blob/master/mainapp/tests.py)

Praful932 commented on mainapp/helpers.py:
```diff
@@ -29,8 +31,8 @@
 
-
 cols = ['original_title', 'authors', 'average_rating', 'image_url', 'book_id']
+priority_list = ['fiction', 'fantasy', 'classics', 'contemporary', 'mystery', 'nonfiction', 'paranormal', 'romance', 'history', 'thriller', 'horror', 'memoir', 'comics', 'biography', 'philosophy', 'science', 'crime', 'psychology', 'christian', 'business', 'poetry', 'music', 'religion', 'manga', 'art', 'spirituality', 'cookbooks', 'travel', 'ebooks', 'sports', 'suspense']
```
Why is this needed?

saaalik replied:
okey, so first off we loop through all the book ids that were given as input and append genres to a list.
that list looks something like this 
```python
['art','art','mystery','romance','romance','fiction','fiction']
```
when this list is passed on to Counter(), it returns something like
```python
{'art':2, 'romance':2, 'fiction':2, 'mystery':1}
```
and the Counter(list).most_common(1) returns the very first item from the dictionary
```
[{'art',2)]
```
but because of this in some cases the order of genre changes after Counter(), resulting in test genre and recommended genre being different, failing the test checks.
So after passing the list to Counter()
```python
{'art':2, 'romance':2, 'fiction':2, 'mystery':1}
```
the genre with highest priority, i.e. the greatest book count is recommended. and `priority_list` stores that order from 0 to n in desending order by

1. Slice out the highest value pairs from the dictionary
```python
most_common_dict = {u: v for u, v in genre_count.items() if v == max_value}
# output {'art':2, 'romance':2, 'fiction':2}
```
2. map out the priority list with index values
```python
index_map = {v: i for i, v in enumerate(priority_list)}
# output {'fiction':0, 'fantasy':1, 'classics':2, 'contemporary':3, ........}
```
3. sorting the dictionary pairs on the index value of a genre
```python
final_list = sorted(most_common_dict.items(), key=lambda pair: index_map[pair[0]])
# output [('fiction',2), ('romance',2), ('art',2)]
```
4. return the highest priority genre
```python
most_common_genre = final_list[0][0]
# output 'fiction'
```

Praful932 replied:
I don't think it is a good practise to hardcode the priority list, instead what you can do to remove the conflict is order the conflict same count genre according to their alphabetical order(take the first one) or according to the count of that genre in the database(take the max one).

saaalik replied:
I have removed the hard coded list and I am implementing a better way to do the same, will make a pr by EOD

Praful932 commented on mainapp/tests.py:
```diff
@@ -209,3 +214,36 @@ def test_user_rated_book_valid(self):
         self.assertEquals(rating.user, self.user)
         self.client.logout()
+
+class MostCommonGenreTestCase(TestCase):
```
We can review this after we are on the same page with the new logic for the recommendations function

Praful932 replied:
Regarding the test don't hardcode the book ids, just take the first 5, then the next 5 and so on as we discussed in discord from the dataframe.
Btw do ping us if you need help

Praful932 replied:
Make a wrapper for all the tests which takes in parameters to avoid duplicate code

Praful932 commented:
Hi @saaalik could you check the tests I had added some comments

saaalik commented:
> Hi @saaalik could you check the tests I had added some comments

I have removed the hard coded priority list as well as the bookids instead randomised their selection and also changed the code such that it finds the book count of the highest frequency books `most_common_dict`, and make a dictionary with key as genre and value as book count, after that the genre with highest book count is selected

Praful932 commented on mainapp/helpers.py:
```diff
@@ -170,26 +170,78 @@ def get_book_dict(bookid_list):
+        # Cut out dictionary containing the highest frequency of genre
+        most_common_dict = {u: v for u, v in genre_count.items() if v == max_value}
+
+        # Sort genre with same frequency based on priority list
```
```suggestion
```

Praful932 reviewed (changes requested):
Update your fork

Praful932 reviewed (changes requested):
Checks are failing :/

saaalik commented:
I've left the percentile input to default, I can't understand its functionality, what should be the ideal value for percentile in `genre_wise` function?

Praful932 commented:
Some checks are failing, you can run the checks on local, see the contributing.md file.
The percentile is just a cutoff, you can think of it as top books that are liked by 85% people, you can leave it at the default value

Praful932 commented on mainapp/tests.py:
```diff
@@ -211,6 +218,61 @@ def test_user_rated_book_valid(self):
+        return most_common_genre
+
+    def test(self):
+        self.template(10, 5, 1)
+        self.template(10, 5, 2)
+        self.template(10, 5, 3)
+        self.template(10, 5, 4)
+        self.template(10, 5, 5)
+        self.template(10, 6, 1)
+        self.template(10, 6, 2)
+        self.template(10, 6, 3)
+        self.template(10, 6, 4)
+        self.template(10, 7, 1)
+        self.template(10, 7, 2)
+        self.template(10, 7, 3)
+        self.template(10, 8, 1)
+        self.template(10, 8, 2)
+        self.template(10, 9, 1)
+        self.template(10, 10, 0)
```
```python
test_cases = [(10, 5, 1), (10, 5, 2), (10, 5, 2), (10, 5, 3), (10, 5, 4),(10, 5, 5), (10, 6, 1), (10, 6, 1), (10, 6, 2), (10, 6, 3), (10, 6, 4),
          (10, 7, 1), (10, 7, 2), (10, 7, 3), (10, 8, 1), (10, 8, 2), (10, 9, 1), (10, 10, 0)]
for tnum, already_slice, bestbookids_slice in test_cases:
    self.template(tnum, already_slice, bestbookids_slice )
```
Also add a docstring to template, so in future, everyone is aware what the params mean

Praful932 reviewed (changes requested):
Added some comments

saaalik commented:
Added the changes, please review

Praful932 reviewed (changes requested):
Changes look fine to me, but the tests are failing :/
Also could you attach the full output of `python manage.py test`

Praful932 commented:
Hi @saaalik could you update the fork, we'll make the changes and merge it

saaalik commented:
I have updated the fork, I am sorry I wasn't able to fix the issue properly, this was my first time contributing to a project of this scale. But I couldn't find what I was doing wrong to get an error in some test cases

Praful932 commented:
No probs

maneprajakta reviewed (approved):
LGTM :rocket:

Praful932 reviewed (approved):
🚀
